### PR TITLE
Mapbox component

### PR DESCRIPTION
**Pull request by jonmattgray**
Add a Map component that wraps `mapbox-gl`.
Map only updates after remounting the component if changing values via the Storybook controls
The component requires an env `STORYBOOK_MAPBOX_TOKEN`

https://user-images.githubusercontent.com/40722025/207622008-9fba65b5-63c0-443b-9c8c-aafa584fd5da.mp4

### CODE REVIEW

File: src/Map/Map.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    },
+    startPosition: {
+      long: args.startPositionLong,
+      lat: args.startPositionLat,
+      zoom: args.startPositionZoom,
+    },
+    style: args.style,
+    sourceJson: args.sourceJson,
+    cluster: args.cluster,
+    clusterOptions: {
```
Comment (n3op2) [inline src/Map/Map.stories.tsx:104]: maybe args could be grouped into table? https://storybook.js.org/docs/react/writing-docs/doc-block-argstable or somehow format a bit nicer? well it's just asthetics, also is this is how many options we need?

File: src/Map/Map.tsx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  const clusterMaxZoom = props.clusterOptions?.clusterMaxZoom || 14
+  const clusterAreaRadius = props.clusterOptions?.clusterAreaRadius || 50
+  const clusterColor = props.clusterOptions?.clusterColor || colors.black
+  const clusterRadius = props.clusterOptions?.clusterRadius || 20
+  const countFont = props.clusterOptions?.countFont || ['Open Sans Regular']
+  const countFontSize = props.clusterOptions?.countFontSize || 14
+  const countFontColor = props.clusterOptions?.countFontColor || colors.white
+  const pointColor = props.clusterOptions?.pointColor || colors.black
+  const pointRadius = props.clusterOptions?.pointRadius || 6
+
```
Comment (n3op2) [inline src/Map/Map.tsx:71]: also this is something i would look into abstracting, but as a round 1 is good

File: src/Map/Map.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      container: mapContainer.current!,
+      style: style,
+      center: [long, lat],
+      zoom: zoom,
+      attributionControl: false,
+    })
+  })
+
+  useEffect(() => {
+    if (!map.current) return // wait for map to initialize
```
Comment (n3op2) [inline src/Map/Map.tsx:84]: just hate empty return statements, i would put null or something

File: src/Map/Map.tsx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      })
+
+      // inspect a cluster on click
+      map.current?.on('click', 'clusters', (e) => {
+        const features = map.current!.queryRenderedFeatures(e.point, {
+          layers: ['clusters'],
+        })
+
+        const clusterId = features[0]?.properties?.cluster_id
+        const source: mapboxgl.GeoJSONSource = map.current!.getSource(
```
Comment (n3op2) [inline src/Map/Map.tsx:138]: nice, not used this operator `key1!.key2`

File: src/Map/Map.tsx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      map.current?.on('mouseenter', 'clusters', () => {
+        map.current!.getCanvas().style.cursor = 'pointer'
+      })
+      map.current?.on('mouseleave', 'clusters', () => {
+        map.current!.getCanvas().style.cursor = ''
+      })
+    })
+  })
+
+  return <Wrapper ref={mapContainer} height={height} width={width}></Wrapper>
```
Comment (n3op2) [inline src/Map/Map.tsx:163]: maybe a jest test against snapshot?

**Review by n3op2 — DISMISSED** [review]

File: src/Map/Map.tsx (hunk 4 above)
Comment (mattdean-digicatapult) [reply src/Map/Map.tsx:138]: what does this do?

File: src/Map/Map.tsx (hunk 3 above)
Comment (jonmattgray) [reply src/Map/Map.tsx]: changed to `return undefined`

File: src/Map/Map.tsx (hunk 2 above)
Comment (jonmattgray) [reply src/Map/Map.tsx]: is this better https://github.com/digicatapult/ui-component-library/pull/26/commits/3fc33c4b0c6a96df4c2472ff33d12ecbb7bdb193?

File: src/Map/Map.tsx (hunk 5 above)
Comment (jonmattgray) [reply src/Map/Map.tsx:213]: I've managed to load in envs fine with `"test": "jest --setupFiles dotenv/config"`, but the Map webgl doesn't load in jest

**Review by mattdean-digicatapult — COMMENTED** [review]
Looking very good to me. Just needs a bit more coverage:

Screenshot 2022-12-14 at 14 42 04

**Comment by mattdean-digicatapult** [comment]
Yes so looking at your Map component my guess is it won’t handle changing props too well. That doesn’t prevent merging but what you probably would need to do is:

* add a sensitivity lists to your effects
* return a cleanup function from your effects to cleanup old data prior to loading in the new data

Unfortunately this will be necessary once we start filtering, but we can merge now so that @n3op2  can get started on the layout.

**Review by mattdean-digicatapult — APPROVED** [review]
There are issues here but my instinct is we merge and then continue to work on it. This won't be a regression

**Pull request merged**